### ISSUE TITLE
Rule proposal: `vue/no-arrow-function`

### ISSUE THREAD

**Issue by armano2** [vuejs/eslint-plugin-vue#152]
**Please describe what the rule should do:**

Warn about potentially dangerous usage of arrow function in component and vue instances

**What category of rule is this? (place an "X" next to just one item)**

`[ ]` Enforces code style
`[x]` Warns about a potential error
`[ ]` Suggests an alternate way of doing something
`[ ]` Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

```js
export default {
  watch: {
    bar: () => {}
  },
  methods: {
     foo: () => {
       this.bar = 2
     }
  }
```

**Comment by mysticatea** [comment]
Thank you for the proposal.

Sounds good to me.
Though, [no-invalid-this](http://eslint.org/docs/rules/no-invalid-this) rule catches this errors since the `this` is invalid.

**Comment by filipalacerda** [comment]
@armano2 @mysticatea 
should this rule be true for all functions inside a vue instance?
`methods`, `computedProps`, `data`, `watch`, etc?

**Comment by SegersIan** [comment]
I guess not everywhere, at least not in `computed`, because if I use `Vuex`
I might map my state like:

```js
{
    computed:{
        ...mapState({
               approvedUsers : (state) => state.userManagement.approvedUsers
         })
    }
}
```

**Comment by michalsnik** [comment]
I also think we should leave it to `no-invalid-this` rule, thus I'm closing this proposal. Feel free to reopen if you still think it's relevant :)

**Comment by sandrosc** [comment]
Well I think the `no-invalid-this` is very far away from the error in the code.
Imagine a user who doesn't know that function types influence `this`. How are they supposed to gain any conclusions from the `no-invalid-this` error message?
